SmartDeviceLink/SDLLifecycleManager.m

```diff
@@ -592,14 +597,25 @@ + (BOOL)sdl_checkNotification:(NSNotification *)notification containsKindOfClass
     return YES;
 }
 
+// this is to make sure that the transition happens on the dedicated queue
+- (void)sdl_transitionToState:(SDLState *)state {
+    if (strcmp(dispatch_queue_get_label(DISPATCH_CURRENT_QUEUE_LABEL), dispatch_queue_get_label(self.lifecycleQueue)) == 0) {
```

Comment:
At first I thought so too, however I realized that it will introduce a rare timing issue.

For instance, let's say the current state is `SDLLifecycleStateReady` and `stop` is called. And just right after the method is called, `transportDidDisconnect` is also called.

The correct sequence should be:
1. The logic of `stop` runs on the queue. State machine transitions to `SDLLifecycleStateUnregistering`.
2. `didEnterStateUnregistering` runs on the queue.
3. The logic of `transportDidDisconnect` runs on the queue. Since the state is now Unregistering, the state machine transitions to `SDLLifecycleStateStopped`.
4. `didEnterStateStopped` runs on the queue.
5. The final state is Stopped.

Now if we use `dispach_async` in `sdl_transitionToState:`, the result can be different:
1. The logic of `stop` runs on the queue, but the transition to `SDLLifecycleStateUnregistering` is deferred with `dispatch_async`.
2. The logic of `transportDidDisconnect` is already scheduled on the queue, and runs. **Since current state is still `SDLLifecycleStateReady`**, a transition to `SDLLifecycleStateReconnecting` is scheduled with `dispatch_async`.
3. Transition to `SDLLifecycleStateUnregistering` happens on the queue.
4. `didEnterStateUnregistering` runs on the queue.
5. Transition to `SDLLifecycleStateReconnecting` happens on the queue.
6. `didEnterStateReconnecting` runs on the queue.
7. The final state is Reconnecting.

I know this isn't a very good example, however I wanted to explain my understanding that the state machine's logic "if XXX then transition to YYY" should be atomic. If it's not, then the logic that runs after it might reference the old state (as you see in step 2) and it might create a timing issue. I would like to avoid that.

Reply:
In that case, should it be run `dispatch_sync` if we're not already on that queue?

Reply:
Hmm, I don't think so. The code that runs before `dispatch_async` should not be related to state machine's work. (The logic of state machine should always run on the dedicated queue.) So I don't see any point to make the call `dispatch_sync` and join the unrelated work.

Reply:
I'm not sure I'm understanding what you're saying. Whether code is written in a `dispatch_async` or `dispatch_sync` block has no bearing on relatedness to the state machine's work or whether or not it's running on a dedicated queue.

`dispatch_async` merely means that the code within the block will run on that queue when resources are available, at the earliest at the beginning of the next run loop, and the current thread **will not** block execution until that occurs. `dispatch_sync` means that the code within the block will run on that queue when resources are available, at the earliest at the beginning of the next run loop, and the current thread **will** block execution until that occurs. There's no difference between where the code is run, it merely alters the execution of the **current thread and queue**. That's why its so dangerous to run `dispatch_sync` on the same queue as the dispatch, because the dispatched code can never run to unblock its own thread.

So I don't think your comment is relevant here. It will be slightly slower to transition `sync` as opposed to `async` since the other queue will be blocked until execution occurs, however, it makes the `transition` call more predictable, since you know the code below it won't execute until the transition is complete, as opposed to an indeterminate time.

Reply:
Thanks for your explanation.
I do understand the difference between `dispatch_sync` and `dispatch_async`, however I didn't have the idea that after `sdl_transitionToState:` is finished, the transition should be completed. Yes, as the method name implies, I think it makes sense to finish the transition within `sdl_transitionToState:` call. If the caller just want to schedule a transition, it can implement like: `dispatch_async(self.lifecycleQueue, ^{ [self sdl_transitionToState:SDLLifecycleStateXxx]; });`

I updated the code with commit 9095a117b9e1bee3c59e267bf7fda629e490eb91.